### PR TITLE
Improve UrlValidation Performance

### PR DESCRIPTION
**Pull request by mikepizzo**
### Issues

*This pull request improves performance of UrlValidation*

### Description

Optimize how UrlValidator eliminates necessary tests.
Make ODataUrlValidator public in order to support caching/re-use of detected optimizations

### Checklist (Uncheck if it is not completed)

- [x] *Test cases added*
- [x] *Build and test with one-click build and test script passed*

### CODE REVIEW

File: src/Microsoft.OData.Core/UrlValidation/ValidationEngine/ODataUrlValidator.cs [hunk 1]
```diff
@@ -144,33 +145,59 @@ private bool ValidateTypeHierarchy(object item, Type itemType, ODataUrlValidatio
             validatedTypes.Add(itemType);
 
             // find any rules defined on an interface defined on the type of the item
-            foreach (Type interfaceType in itemType.GetInterfaces())
+            // GetInterfaces() returns derived interfaces, so only do this at the top level
+            if (includeInterfaces)
             {
-                if (validatedTypes.Add(interfaceType) && !this.unusedTypes.ContainsKey(interfaceType))
+                foreach (Type interfaceType in GetInterfaces(itemType))
                 {
```
Comment (Sreejithpin) [inline src/Microsoft.OData.Core/UrlValidation/ValidationEngine/ODataUrlValidator.cs:152]: Can we use for loops as its array

File: src/Microsoft.OData.Core/UrlValidation/ValidationEngine/ODataUrlValidator.cs [hunk 2]
```diff
@@ -411,26 +438,29 @@ private void ValidateImpliedProperties(IEdmType segmentType, SelectExpandClause
         private void ValidateProperties(IEdmType edmType, ODataUrlValidationContext context)
```
Comment (Sreejithpin) [inline src/Microsoft.OData.Core/UrlValidation/ValidationEngine/ODataUrlValidator.cs:438]: Could this be called multiple times for same edmtype? if so can this be cahced

File: test/FunctionalTests/Microsoft.OData.Core.Tests/UrlValidation/ODataUrlValidatorTests.cs [hunk 3]
```diff
@@ -55,17 +55,121 @@ public static void FailedParse()
...
+            Assert.Collection<ODataUrlValidationMessage>(errors, (error) =>
+            {
+                Assert.Equal("dummyCode", error.MessageCode);
+                Assert.Equal("dummyMessage", error.Message);
+                Assert.Equal(OData.UriParser.Validation.Severity.Warning, error.Severity);
+            });
+        }
+
+        [Fact]
+        public static void PassingValdiatorSucceeds()
```
Comment (xuzhg) [inline test/FunctionalTests/Microsoft.OData.Core.Tests/UrlValidation/ODataUrlValidatorTests.cs:97]: >Valdiator [](start = 34, length = 9)

typo

**Review by xuzhg — APPROVED** [review]
:shipit:

File: src/Microsoft.OData.Core/UrlValidation/ValidationEngine/ODataUrlValidator.cs (hunk 1 above)
Comment (mikepizzo) [reply src/Microsoft.OData.Core/UrlValidation/ValidationEngine/ODataUrlValidator.cs:152]: Good call; I like it.

---
In reply to: [518474457](https://github.com/OData/odata.net/pull/1917#discussion_r518474457) [](ancestors = 518474457)

File: src/Microsoft.OData.Core/UrlValidation/ValidationEngine/ODataUrlValidator.cs (hunk 2 above)
Comment (mikepizzo) [reply src/Microsoft.OData.Core/UrlValidation/ValidationEngine/ODataUrlValidator.cs:438]: there's not really anything to cache (the method looks for and executes any rules that are applied to the particular type, and the rules add any exceptions to the validation context). However, we go one better, and on line 441 we skip validation if we've already run validation on this type. 

---
In reply to: [518476690](https://github.com/OData/odata.net/pull/1917#discussion_r518476690) [](ancestors = 518476690)

**Pull request merged**